src/analyser.cpp

```diff
@@ -0,0 +1,1509 @@
+                                         const AnalyserInternalEquationPtr &equation)
+{
+    // Create the AST, if needed.
+
+    if (ast.get() == nullptr) {
+        ast.reset(new AnalyserEquationAst {});
+    }
+
+    // Basic content elements.
+
+    if (node->isMathmlElement("apply")) {
+        // We may have 2, 3 or more child nodes, e.g.
+        //
+        //                 +--------+
+        //                 |   +    |
+        //        "+a" ==> |  / \   |
+        //                 | a  nil |
+        //                 +--------+
+        //
+        //                 +-------+
+        //                 |   +   |
+        //       "a+b" ==> |  / \  |
+        //                 | a   b |
+        //                 +-------+
+        //
+        //                 +-------------+
+        //                 |   +         |
+        //                 |  / \        |
+        //                 | a   +       |
+        //                 |    / \      |
+        // "a+b+c+d+e" ==> |   b   +     |
+        //                 |      / \    |
+        //                 |     c   +   |
+        //                 |        / \  |
+        //                 |       d   e |
+        //                 +-------------+
+
+        auto childCount = mathmlChildCount(node);
+
+        processNode(mathmlChildNode(node, 0), ast, astParent, component, equation);
+        processNode(mathmlChildNode(node, 1), ast->mPimpl->mLeftChild, ast, component, equation);
+
+        if (childCount >= 3) {
+            AnalyserEquationAstPtr astRightChild;
+            AnalyserEquationAstPtr tempAst;
+
+            processNode(mathmlChildNode(node, childCount - 1), astRightChild, nullptr, component, equation);
+
+            for (auto i = childCount - 2; i > 1; --i) {
+                tempAst = AnalyserEquationAst::create();
+
+                processNode(mathmlChildNode(node, 0), tempAst, nullptr, component, equation);
+                processNode(mathmlChildNode(node, i), tempAst->mPimpl->mLeftChild, tempAst, component, equation);
+
+                astRightChild->mPimpl->mParent = tempAst;
+
+                tempAst->mPimpl->mRightChild = astRightChild;
+                astRightChild = tempAst;
+            }
+
+            if (astRightChild != nullptr) {
+                astRightChild->mPimpl->mParent = ast;
+            }
+
+            ast->mPimpl->mRightChild = astRightChild;
+        }
+
+        // Assignment, and relational and logical operators.
+
+    } else if (node->isMathmlElement("eq")) {
+        // This element is used both to describe "a = b" and "a == b". We can
+        // distinguish between the two by checking its grand-parent. If it's a
+        // "math" element then it means that it is used to describe "a = b"
+        // otherwise it is used to describe "a == b". In the former case, there
+        // is nothing more we need to do since `ast` is already of
+        // AnalyserEquationAst::Type::ASSIGNMENT type.
```

Comment:
Are equations like `a=b=c` allowed?  If so, checking the grandparent of `c` will make you think it's an equality not an assignment?

Reply:
I don't think `a=b=c` is allowed. Is that correct @nickerso?

Reply:
Ping @nickerso.

Reply:
`eq` is the equality operator, there is no assignment operator. `a=b=c` is perfectly valid.

Reply:
I thought this was something we do allow in CellML but don't handle.

Reply:
Ok, next question: do we want to support this kind of MathML statement? Clearly, we don't at this stage. In the same way, we don't currently support something like `b+c = a` where both `b` and `c` are known. Maybe those two things should be addressed in issue #688?

Reply:
It's valid but not something that the Analyser can interpret?  There will be loads of examples of those, I'd imagine?

Reply:
Indeed, the `Analyser` cannot interpret this, at this point. For now, I have updated issue #688 to deal with this, although we may decide not to in the end. We will see.

Otherwise, I am not aware of any CellML file that has things like `a = b = c`. (I wonder what the CellML API does about it...?)

Reply:
Resolving this conversation since this will be addressed in some way or another (i.e. will be supported or not, but at least it will handled) in issue #688.